posts/frangeris/04.upgradeable-contracts.md

```diff
@@ -0,0 +1,228 @@
+  uint private fee;
+
+  function initialize(uint _fee) public initializer {
+    fee = _fee;
+    __Ownable_init();
+  }
+
+  function getFee() public view returns (uint) {
+    return fee;
+  }
+}
+```
+
+This contract is very simple, it has a `fee` variable that can be set by the owner of the contract. This is the contract that we are going to be upgrading.
+
+To deploy the contract we need to use the `deployProxy` function from the `@openzeppelin/hardhat-upgrades` plugin. This function will deploy the proxy contract and initialize it with the `initialize` function from the `Gateway` contract.
+
+To keep things simple and visible (in case you cant deploy), we will use [Sepolia testnet](https://sepolia.dev/) blockchain and [Hardhat](https://hardhat.org/) as development environment, so to configure hardhat to use sepolia we need to add the following configuration to the `hardhat.config.js` file:
+
+```javascript
+  networks: {
+    sepolia: {
+      url: 'https://rpc.sepolia.org/',
+
+      // take the private key from the first account in the ganache-accounts
+      accounts: [process.env.ACCOUNT_PRIVATE_KEY!],
+    },
+  },
+```
+
+Then we can tell hardhat to run the deployment script while specifying the network to use:
+
+```bash
+npx hardhat run scripts/deploy.js --network sepolia
+```
+
+![Output of deploy script](./assets/04.image.2.png)
+
+You should be able to see the same transaction on [Sepolia Block Explorer](https://sepolia.etherscan.io/address/0x10Ccb5C22C6Ccc78F6B5dfbCB36Ea13bF60d2Ecf), now let's take a look at the transactions details (lookup your wallet address in etherscan), `3` contracts were created:
+
+1. `TransparentUpgradeableProxy`: This is the proxy contract that will be used to interact with the Gateway contract, [see it here](https://sepolia.etherscan.io/address/0xe7197e1d5f60ee3745a828a4ceadaa0fbfff886d#code).
+2. `ProxyAdmin`: This contract is used to manage the proxy contracts (admin / upgrades), [see it here](https://sepolia.etherscan.io/address/0xe754bdb16e0fc27d5d405432c523099c2c9050b2#code).
+3. `Gateway`: This is the logic contract that will be used by `TransparentUpgradeableProxy`, [see it here](https://sepolia.etherscan.io/address/0x739af28dfecd11067f5be66f56f3fb37403c496f#code).
+
+> Note that Gateway contract code is not available as the others, only the bytecode is shown, this is because it was deployed by the proxy, to be able to see the real code you can verify and publish it via hardhat.
+
+Any future interaction with the `Gateway` contract will be done via the `TransparentUpgradeableProxy` contract, so let's try to interact with it. To do so, let's use `hardhat console` to interact with the blockchain:
+
+```bash
+npx hardhat console --network sepolia
+```
+
+Why specifying the network? Because we need to tell hardhat which network to use to interact with the blockchain, in this case we are using the `sepolia` network, also we will use `ethersjs` to interact with the deployed contract.
```

Comment:
```suggestion
Why specify the network? Because we need to tell Hardhat which network to use to interact with the blockchain. In this case we are using the `sepolia` network and we will also use `ethersjs` to interact with the deployed contract.
```